### PR TITLE
Scaffold

### PR DESCRIPTION
**Pull request by fortmarek**
Resolves https://github.com/tuist/tuist/issues/813

### Short description 📝

This is a WIP PR that implements [#813](https://github.com/tuist/tuist/issues/813). Creating new frameworks, features is not really extensible and is basically constrained to `tuist init` and manually adding needed files - which is not the user experience we want!

### Solution 📦

Scaffold solves the aforementioned issues by bringing templates and where `tuist scaffold name_of_feature` will create all the files from the template - either from `~/.tuist/Versions/x.y.z/Templates` or from the current project's `Tuist/Templates` directory.

### Implementation 👩‍💻👨‍💻

- [x] Add Template to `ProjectDescription`
- [x] Add Scaffold command
- [x] Add `tuist scaffold --list` to list available templates
- [x] Parse user input plus the given template and generate the needed files
- [x] Define Tuist built-in templates
- [x] Copy Tuist templates upon installation (similarly to `ProjectDescription` framework)
- [x] Load `Template.swift` in `tuist edit`
- [x] Add option to generate file elsewhere
- [x] Add tests
- [x] Add documentation
- [x] Add changelog

### Current State

To get the prototype working, simply add a `Template.swift` file from below to your current project's `Tuist/Templates` directory and run `tuist scaffold custom --attributes name="My custom string"`

This should generate a directory `custom` with a file `custom/custom.swift` that says:
`// this is some My custom string content`

`Template.swift` file:
```swift
import ProjectDescription

let nameArgument: Attribute = .required("name")

let testContents = """
// this is some \(nameArgument) content

"""

let template = Template(
    description: "Custom template",
    arguments: [nameArgument],
    files: [File(path: "custom/custom.swift", contents: testContents)],
    directories: ["custom"]
)
```

I'd appreciate any early feedback, especially to the manifest and `--attributes` solution as that is something that has not been discussed in the prior issue.

The drawback of the current solution is that the contents of the file have to be all declared in the single `Template.swift` - but it should be possible to load the contents of the file from the current `Templates` directory.

Also note this is an early prototype and there is a lot of rough edges.

### CODE REVIEW

File: Sources/ProjectDescription/Template.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
+public struct Template: Codable {
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template.swift]: Since this is part of the public interface, I'd document the attributes and the methods. That way the user can get inline documentation while editing the template on Xcode.

Comment (pepibumur) [inline Sources/ProjectDescription/Template.swift]: I'd import Foundation at the top

File: Sources/ProjectDescription/Template.swift [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    public let description: String
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template.swift]: Did you check if the templates are part of the project generated with the `tuist edit` command?

File: Sources/ProjectDescription/Template.swift [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public let path: String
+    public let contents: String
+    
+    public init(path: String, contents: String) {
+        self.path = path
+        self.contents = contents
+    }
+}
+
+public enum Attribute: Codable {
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template.swift]: Since this is specific to the `Template` entity, what about placing it under it? E.g.: `Template.File`

File: Sources/ProjectDescription/Template.swift [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+public extension String.StringInterpolation {
+    mutating func appendInterpolation(_ value: Attribute) {
+        switch value {
+        case let .required(name), let .optional(name, default: _):
+            appendInterpolation("{{ \(name) }}")
+        }
+    }
+}
+
+public struct File: Codable {
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template.swift]: Similar comment here in regards to placing the definition of `File` under `Template`.

File: Sources/ProjectDescription/Template.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                directories: [String] = []) {
+        self.description = description
+        self.attributes = arguments
+        self.files = files
+        self.directories = directories
+        dumpIfNeeded(self)
+    }
+}
+
+public extension String.StringInterpolation {
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template.swift]: Do you have an example of what this extension is for? I can't see its use case from the implementation.

File: Sources/ProjectDescription/Template.swift (hunk 4 above)
Comment (fortmarek) [reply Sources/ProjectDescription/Template.swift]: Yes, I will fill in the documentation once we are settled on the `template` manifest.

File: Sources/TuistKit/Commands/ScaffoldCommand.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import Foundation
+import TuistSupport
+import TuistLoader
+import SPMUtility
+import Basic
+
+// swiftlint:disable:next type_body_length
+class ScaffoldCommand: NSObject, Command {
```
Comment (pepibumur) [inline Sources/TuistKit/Commands/ScaffoldCommand.swift]: Beautifully implemented @fortmarek

File: Sources/ProjectDescription/Template.swift (hunk 5 above)
Comment (fortmarek) [reply Sources/ProjectDescription/Template.swift]:
```swift
let testContents = """
// this is some \(nameArgument) content
"""
```
-> I need to encode and then decode contents of the file - and that means encoding and decoding the attribute; this turns it to `{{ attribute }}` for which I can then later fill in the value that the user inputs.

File: Sources/ProjectDescription/Template.swift (hunk 2 above)
Comment (fortmarek) [reply Sources/ProjectDescription/Template.swift]: At the moment it does not seem so - but it probably should if I am not mistaken? 🙂 If so, I'll add it to the implementation check list.

File: Sources/ProjectDescription/Template.swift (hunk 3 above)
Comment (fortmarek) [reply Sources/ProjectDescription/Template.swift]: Yeah, that's a good idea, will do 👍

File: Sources/ProjectDescription/Template.swift (hunk 5 above)
Comment (pepibumur) [reply Sources/ProjectDescription/Template.swift]: What about using the ExpressibleByStringInterpolation protocol like we are doing [here](https://github.com/tuist/tuist/blob/master/Sources/TuistSupport/Utils/Printer.swift#L21)? We [follow](https://github.com/tuist/tuist/blob/master/Sources/ProjectDescription/TuistConfig.swift#L9) a similar approach to define the naming convention for projects.

Comment (fortmarek) [reply Sources/ProjectDescription/Template.swift]: Nice, I like that more, I will rewrite it 👍

Comment (fortmarek) [reply Sources/ProjectDescription/Template.swift]: So, I actually don't see the benefits there, would you mind to give a little bit of explaining? Why declare a struct such as [TemplateString](https://github.com/tuist/tuist/blob/master/Sources/ProjectDescription/TemplateString.swift) when all I need is custom `appendInterpolation` method? Seems to me like an unnecessary overhead.

File: Sources/TuistKit/Commands/ScaffoldCommand.swift (hunk 6 above)
Comment (fortmarek) [reply Sources/TuistKit/Commands/ScaffoldCommand.swift]: Thanks 🙏

**Comment by fortmarek** [comment]
@ollieatkinson I have added a fixture, you can find it at `fixtures/ios_app_with_templates`
To generate the current custom template that is located there, run this in the directory mentioned above:
```
tuist scaffold custom --attributes --name MyScaffold --platform ios
```

I have also added a built-in template that you can find in `Templates` directory - unfortunately, right now you have to copy it in your `DerivedData Products` folder (eg mine is `Users/marekfort/Library/Developer/Xcode/DerivedData/tuist-ckenurlpkadgjzgyvspugoglsmsu/Build/Products/Debug/Templates`), so `ResourceLocator` can find it. The current solution will work for the end user since on install I can copy the `Templates` directory to the same location as binary, but I am not sure how to do this when running from Xcode (frameworks such as `ProjectDescription` get copied alongside with the binary, but this can not be done for resource files such as `Templates`)

You can see that I have added two new concepts:
1) `.generated` files that can contain a bit more elaborate logic than just inserting a string from user input (you can see that in the built-in `default` Template)
2) Template description helpers to share code between templates (this is very similar to project description helpers with a few differences)

**Comment by fortmarek** [comment]
I feel like the feature is now ready for more thorough review!

Only thing that I still feel not great about is the fact that you have to add `--attributes` eg:
`tuist scaffold template --attributes --name Name`

Unfortunately, I was not able to go around that.

To learn more about how this feature is intended to be used you can take a look at the [documentation](https://github.com/tuist/tuist/blob/scaffold/website/markdown/docs/usage/commands.mdx)

File: fixtures/ios_app_with_templates/Tuist/Templates/custom/Template.swift [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+// this is test \(nameAttribute) content
+"""
+
+let template = Template(
+    description: "Custom template",
+    attributes: [
+        nameAttribute,
+        platformAttribute
+    ],
+    files: [
```
Comment (pepibumur) [inline fixtures/ios_app_with_templates/Tuist/Templates/custom/Template.swift]: I looked at this file before the code and the documentation to see if I'd be able to understand it, and these are the things that popped in my mind:

- What do `directories` mean? 
- Why is the file `static` when it's dynamically generated when scaffolding?
- What does `generateFilePath` mean?
- Why `custom_dir` appears both in the `files` and also in the `directories` attribute?

File: Rakefile [hunk 8]
```diff
@@ -101,7 +101,8 @@ def package
     system(
       "zip", "-q", "-r", "--symlinks",
       "tuist.zip", "tuist",
-      "ProjectDescription.swiftmodule", "ProjectDescription.swiftdoc", "libProjectDescription.dylib", "ProjectDescription.swiftinterface"
+      "ProjectDescription.swiftmodule", "ProjectDescription.swiftdoc", "libProjectDescription.dylib", "ProjectDescription.swiftinterface",
+      "../../Templates"
```
Comment (pepibumur) [inline Rakefile]: I'd avoid paths relative to the path `rake` is executed from. You can replace `../../Templates` with `File.expand_path("Templates", __dir__)`

File: Sources/ProjectDescription/Template/Template.swift [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        self.files = files
+        self.directories = directories
+        dumpIfNeeded(self)
+    }
+
+    /// Enum containing information about how to generate file
+    public enum Contents: Codable, Equatable {
+        /// Static Contents is defined in `Template.swift` and contains a simple `String`
+        /// Can not contain any additional logic apart from plain `String` from `arguments`
+        case `static`(String)
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template/Template.swift]: What about calling this one `string`?

File: Sources/ProjectDescription/Template/Template.swift [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        /// Generated content is defined in a different file from `Template.swift`
+        /// Can contain additional logic and anything that is defined in `ProjectDescriptionHelpers`
+        case generated(String)
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template/Template.swift]: What about calling this one `.file`?

File: Sources/TuistEnvKit/Installer/BuildCopier.swift [hunk 11]
```diff
@@ -12,6 +12,7 @@ class BuildCopier: BuildCopying {
     /// Files that should be copied (if they exist).
     static let files: [String] = [
         "tuist",
+        Constants.templatesDirectoryName,
```
Comment (pepibumur) [inline Sources/TuistEnvKit/Installer/BuildCopier.swift]: If the user tries to install a version of Tuist that doesn't include `/Templates` the `copy` method below will fail because `/bin/cp` will exit unsuccessfully. You'll need to handle that scenario gracefully:

```swift
class BuildCopier {
  static let files: [(path: String, optional Bool)] = [
    (path: Constants.templatesDirectoryName, optional: true)
  ]

}
```

File: Sources/TuistEnvKit/Installer/Installer.swift [hunk 12]
```diff
@@ -180,6 +180,9 @@ final class Installer: Installing {
             }
             try FileHandler.shared.createFolder(installationDirectory)
 
+            try FileHandler.shared.copy(from: temporaryDirectory.path.appending(component: Constants.templatesDirectoryName),
```
Comment (pepibumur) [inline Sources/TuistEnvKit/Installer/Installer.swift]: Similar comment here. If the user tries to install an old version of Tuist that doesn't contain the templates directory. This command will fail.

File: Sources/ProjectDescription/Template/AttributeHelpers.swift [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        }
+    }
+}
+
+/// Returns value for `ParsedAttribute` of `name` from user input
+/// - Parameters:
+///     - name: Name of `ParsedAttribute`
+///     - arguments: Arguments to get attribute from (usually you should lead this at default)
+/// - Returns: Value of `ParsedAttribute`
+public func getAttribute(for name: String, arguments: [String] = CommandLine.arguments) throws -> String {
```
Comment (pepibumur) [inline Sources/ProjectDescription/Template/AttributeHelpers.swift]: Models in the `ProjectDescription` framework are intentionally designed to be dumb: they contain a constructor, attributes, and the logic to encode and decode them. All the business logic lives on the `TuistLoader` and the `TuistGenerator` side.

This implementation is not consistent with that. Models parse information passed as shell arguments, that they are then using to substitute variables in the template. I'd suggest changing the approach.

What Tuist gets from the `ProjectDescription` is just a description of the template, being the template metadata with references to files or content that might contain attributes. To be able to substitute those variables, we need them to have a special format within the template. That's something up to us to define, but it could be something along the lines of:

```swift
// Swift file

class {{name:string:true}}Interactor {

}
```

Or easier, we can follow [Stencil's template system](https://github.com/stencilproject/Stencil). After loading the template in memory, we can parse the arguments and do the substitution of the variables. If there are missing variables or mismatching of types, we can fail and exit with an error. If the substitution runs successfully, then we pass the template down to the scaffolding utility to generate the template.

What do you think @fortmarek ?

Comment (fortmarek) [reply Sources/ProjectDescription/Template/AttributeHelpers.swift]: Thanks for the thorough comment @pepibumur 🙂 

This bit of logic was needed in order to be able to define custom templates in Swift - which has its benefits like not forcing developers to learn a template language (though I realize that Stencil is a de facto standard in this regard) and as in all other Tuist manifests the ability to leverage Swift as a language with variables and type-safety. Besides that there is for example a possibility for the user to define custom errors right in the template - shared with `ProjectDescriptionHelpers` across the templates.

That being said, I agree that keeping `ProjectDescription` simple as possible is definitely a high priority. 

If you think that trying to work around it is worth the effort, I can have another look if there is a way how to get rid of the business logic here (or at least to the point where it is simple as, say, a decoding). Otherwise, we can definitely fall back to Stencil, I have tried to take a different approach, but I am not at all saying it is definitely the right one.

File: fixtures/ios_app_with_templates/Tuist/Templates/custom/Template.swift (hunk 7 above)
Comment (fortmarek) [reply fixtures/ios_app_with_templates/Tuist/Templates/custom/Template.swift]: `directories` is a list of directories that should be generated - this has two implications: creating directories is an explicit operation and is not hidden in a file structure of the files. That means that if user wants to quickly see what directories will be generated, this is the one place they can look at. As a side benefit, users can create empty directories (though it is arguable how useful it is ... ) But we can also derive the generated directories from the files, which would make the setup a little bit simpler.
To summarize, do you feel like that property should be renamed or dropped altogether?

**Comment by fortmarek** [comment]
Closing this PR as it will be split into multiple PRs to be more digestible.

Currently, we plan to split it into these chunks:
1. **ProjectDescription Models:** Models in the project description framework. You can find it [here](https://github.com/tuist/tuist/pull/1082) already.
2. **Model loading and variable substitution:** The logic that loads the template, and substitutes the variables reading values from the CLI. 
3. **Generation of the files from the template:** The actual business logic of the feature that takes the template and generates the files.
4. **Built-in templates:** Defining templates that we want to have built-in. That also includes logic to include those during installation.

File: Sources/ProjectDescription/Template/AttributeHelpers.swift (hunk 13 above)
Comment (pepibumur) [reply Sources/ProjectDescription/Template/AttributeHelpers.swift]: Hey @fortmarek 

> This bit of logic was needed in order to be able to define custom templates in Swift - which has its benefits like not forcing developers to learn a template language (though I realize that Stencil is a de facto standard in this regard) and as in all other Tuist manifests the ability to leverage Swift as a language with variables and type-safety.

I see where your rationale came from. I think it's fine adopting a template system like Stencil. It's very standard and easy to learn. The benefit of using it is not just simplifying the models in the `ProjectDescription` framework, but also being able to use the same template system for both, templates defined in a file or in the `Template.swift` file.

> That being said, I agree that keeping ProjectDescription simple as possible is definitely a high priority.

Not a priority, but I think is good to stay consistent with how we've been doing things. If at some point it proves not to be the best decision going forward, we can reconsider it. It's been working well so far, and I think this feature can be well implemented following the same approach.

> If you think that trying to work around it is worth the effort, I can have another look if there is a way how to get rid of the business logic here (or at least to the point where it is simple as, say, a decoding). Otherwise, we can definitely fall back to Stencil, I have tried to take a different approach, but I am not at all saying it is definitely the right one.

I'd suggest towards Stencil and doing the replacement of the variables after the template has been loaded.

**Comment by pepibumur** [comment]
Thanks for splitting it up and being so comprehensive with the feedback @fortmarek. I'm very sorry I could not bring my concerns earlier because that would have said us some time. Having said that, if you'd like me to help with the Stencil adoption, I'll be glad to give you a hand. I'm very excited for this feature and I think a lot of users that are already using Tuist will love it too.
